src/dispatch.c

```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+
+        // --- Normal user message delivery. ---
+
+        // Is this a subscription that can timeout?
+        if (!sub->draining && (sub->timeout != 0))
+        {
+            // Prevent the timer from posting a timeout synthetic message.
+            sub->timeoutSuspended = true;
+
+            // If we are dealing with the last pending message for this sub,
+            // we will reset the timer after the user callback returns.
+            if (sub->ownDispatcher.queue.msgs == 0)
+                timerNeedReset = true;
+        }
+
+        nats_unlockDispatcher(d);
+
+        // If we are fetching, see if we need to ask the server for more.
+        if (fetch != NULL)
+            js_maybeFetchMore(sub, fetch);
+
+        if (!overLimit)
+            (*messageCB)(nc, sub, msg, messageClosure);
+        else
+            natsMsg_Destroy(msg);
+
+        if (fcReply != NULL)
+        {
+            natsConnection_Publish(nc, fcReply, NULL, 0);
+            NATS_FREE(fcReply);
+        }
+
+        if (lastMessageInFetch)
+        {
+            if (fetch->completeCB != NULL)
+                fetch->completeCB(nc, sub, fetchStatus, fetch->completeCBClosure);
+        }
+
+        // If we have reached the sub's message max, we need to remove
+        // the sub.
+        if (lastMessageInSub)
+        {
+            // Call this blindly, it will be a no-op if the subscription
+            // was not draining.
+            natsSub_setDrainCompleteState(sub);
+            natsConn_removeSubscription(nc, sub);
+        }
+
+        nats_lockDispatcher(d);
+
+        // Check if timer need to be reset for subscriptions that can timeout.
+        if (!sub->closed && (sub->timeout != 0) && timerNeedReset)
+        {
+            timerNeedReset = false;
+
+            // Do this only on timer reset instead of after each return
+            // from callback. The reason is that if there are still pending
+            // messages for this subscription (this is the case otherwise
+            // timerNeedReset would be false), we should prevent
+            // the subscription to timeout anyway.
+            sub->timeoutSuspended = false;
+
+            // Reset the timer to fire in `timeout` from now.
+            natsTimer_Reset(sub->timeoutTimer, sub->timeout);
+        }
+    }
+
+    nats_destroyQueuedMessages(&d->queue);
+    nats_unlockDispatcher(d);
+
+    natsLib_Release();
+}
+
+// Thread main function for a dedicated dispatcher.
+void nats_dispatchThreadDedicated(void *arg)
+{
+    natsSubscription *sub = (natsSubscription *)arg;
+    bool rmSub = false;
+
+    // These are set at sub creation time and never change, no need to lock.
+    natsConnection *nc = sub->conn;
+    natsMsgHandler messageCB = sub->msgCb;
+    void *messageClosure = sub->msgCbClosure;
+    natsOnCompleteCB completeCB = NULL;
+    void *completeCBClosure = NULL;
+
+    // This just serves as a barrier for the creation of this thread.
+    natsConn_Lock(nc);
```

Comment:
I wonder if instead of having those we should have the sub lock/unlock when getting the fields above, even though we say that they are immutable.

Reply:
It's not even the locking, no need to refresh them every iteration of the loop; and since they are already set when the thread is started and immutable after, I grouped them before the loop. Change the comment?

Reply:
Actually, I added that back [here](https://github.com/nats-io/nats.c/commit/4b64e7a3b41086b7442601d867815159b60bc6be).

Maybe that was a different time and trying to protect against early run of the dispatch code before the subscription was fully configured? Could it be that there was a chance that - since the proto is already sent - messages would be already coming and possibly user callback invoked even before the `natsConnection_Subscribe()` call returned?
So either let's leave it, or make sure that it is safe to remove.

Reply:
Sorry I misunderstood your original comment. 1/5 Let's just leave it here as is, it's not expensive and been there for some time. I wonder if there needs to be an equivalent barrier (and what it would be) in the pool version, if the condition you described can happen there.